lib/localisation/src/filtering/kalman_filter.rs

```diff
@@ -0,0 +1,164 @@
+use nalgebra::{Matrix2, Vector1, Vector2};
+/// Kalman filter implementation for cart for sensor fusion.
+/// Recursively estimates the state from a series of nois measurements.
+///
+/// Uses keyence, optical flow and accelerometer data.
+///
+/// Control input: Accelerometer data
+/// Measurement: Optical flow data, keyence stripe counter
+pub struct KalmanFilter {
+    /// Current state estimate (2x1)
+    state: Vector2<f64>,
+    /// Current error covariance (2x2)
+    covariance: Matrix2<f64>,
+    /// State transition matrix (2x2)
+    transition_matrix: Matrix2<f64>,
+    /// Control matrix (2x1)
+    control_matrix: Vector2<f64>,
+    /// Observation matrix (1x2)
+    observation_matrix: Matrix2<f64>,
+    /// Process noise covariance (2x2)
+    process_noise: Matrix2<f64>,
+    /// Measurement noise covariance (1x1)
+    measurement_noise: Matrix2<f64>,
+}
+
+impl KalmanFilter {
+    pub fn new(
+        initial_state: Vector2<f64>,
+        initial_covariance: Matrix2<f64>,
+        transition_matrix: Matrix2<f64>,
+        control_matrix: Vector2<f64>,
+        observation_matrix: Matrix2<f64>,
+        process_noise: Matrix2<f64>,
+        measurement_noise: Matrix2<f64>,
+    ) -> Self {
+        KalmanFilter {
+            state: initial_state,
+            covariance: initial_covariance,
+            transition_matrix,
+            observation_matrix,
+            control_matrix,
+            process_noise,
+            measurement_noise,
+        }
+    }
+
+    /// Predict step
+    /// Predicts the next state of the system, using the accelerometer data as control input.
+    /// Predicts the next state covariance.
+    pub fn predict(&mut self, control_input: &Vector1<f64>) {
+        // x_k = F * x_k-1 + B * u_k
+        self.state = self.transition_matrix * self.state + self.control_matrix * control_input;
+
+        // P_k = F * P_k-1 * F^T + Q
+        self.covariance =
+            self.transition_matrix * self.covariance * self.transition_matrix.transpose()
+                + self.process_noise;
+    }
+
+    /// Update step: Corrects the state estimate based on the measurement.
+    /// Uses the stripe counter (displacement) and optical flow data (velocity).
+    pub fn update(&mut self, measurement: &Vector2<f64>) {
+        // y_k = z_k - H * x_k
+        let innovation = measurement - self.observation_matrix * self.state;
+
+        // S = H * P_k * H^T + R
+        let innovation_covariance =
+            self.observation_matrix * self.covariance * self.observation_matrix.transpose()
+                + self.measurement_noise;
+
+        // Calculate the inverse of the innovation covariance
+
+        let a = innovation_covariance[(0, 0)];
+        let b = innovation_covariance[(0, 1)];
+        let c = innovation_covariance[(1, 0)];
+        let d = innovation_covariance[(1, 1)];
+
+        let determinant = a * d - b * c;
+
+        let innovation_covariance_inv = Matrix2::new(d, -b, -c, a) / determinant;
```

Comment:
is the determinant definitely never going to be 0? if it is 0 this panics, if you can guarantee it isn't then there isn't a need to manually do the inverse instead of using the lib function + `unwrap()` (unless there is and i'm unaware)

Reply:
add a comment that explains why the determinant will never be 0 and thus the inverse cannot fail